### ISSUE TITLE
audio devices missing

### ISSUE THREAD

**Issue by adminy** [Aylur/astal#23]
`wpctl status`
```
PipeWire 'pipewire-0' [1.2.3, adminy@box, cookie:2521344676]
 └─ Clients:
        34. WirePlumber                         [1.2.3, adminy@box, pid:1533]
        35. pipewire                            [1.2.3, adminy@box, pid:1555]
        42. .xdg-desktop-portal-hyprland-wrapped [1.2.3, adminy@box, pid:1679]
        49. WirePlumber [export]                [1.2.3, adminy@box, pid:1533]
        50. com.github.Aylur.ags mixer control  [1.2.3, adminy@box, pid:1538]
        78. .easyeffects-wrapped                [1.2.3, adminy@box, pid:2452]
       118. Chromium input                      [1.2.3, adminy@box, pid:57410]
       147. wpctl                               [1.2.3, adminy@box, pid:767309]
       178. Firefox                             [1.2.3, adminy@box, pid:745443]

Audio
 ├─ Devices:
 │      51. GENERAL WEBCAM                      [alsa]
 │      52. K38                                 [alsa]
 │      53. GA102 High Definition Audio Controller [alsa]
 │      54. Starship/Matisse HD Audio Controller [alsa]
 │     160. Blue Microphones                    [alsa]
 │
 ├─ Sinks:
 │      60. GA102 High Definition Audio Controller Digital Stereo (HDMI) [vol: 0.95]
 │      61. Starship/Matisse HD Audio Controller Digital Stereo (IEC958) [vol: 3.05]
 │      79. Easy Effects Sink                   [vol: 1.00]
 │  *  155. Blue Microphones Analog Stereo      [vol: 0.46]
 │
 ├─ Sources:
 │      58. GENERAL WEBCAM Mono                 [vol: 1.07]
 │      59. K38 Mono                            [vol: 0.10]
 │      62. Starship/Matisse HD Audio Controller Analog Stereo [vol: 1.00]
 │      80. Easy Effects Source                 [vol: 1.00]
 │  *  159. Blue Microphones Analog Stereo      [vol: 1.00]
 │
 ├─ Filters:
 │
 └─ Streams:
       210. Firefox
            148. output_FR       > Easy Effects Sink:playback_FR	[active]
            208. output_FL       > Easy Effects Sink:playback_FL	[active]

Video
 ├─ Devices:
 │      45. GENERAL WEBCAM                      [v4l2]
 │      47. GENERAL WEBCAM                      [v4l2]
 │      63. GENERAL WEBCAM: GENERAL WEBCAM      [libcamera]
 │
 ├─ Sinks:
 │
 ├─ Sources:
 │  *   64. GENERAL WEBCAM (V4L2)
 │
 ├─ Filters:
 │
 └─ Streams:

Settings
 └─ Default Configured Devices:
         0. Audio/Sink    alsa_output.usb-Generic_Blue_Microphones_2036BAB08KK8-00.analog-stereo
```

```typescript
const { audio } = Wp.get_default()!
const micMenu = Menu(audio.microphones || [])
const speakersMenu = Menu(audio.speakers || [])
```
but of course both mics and speakers are null.
Also .get_microphones() and .get_speakers() returns null.

Stranger yet is that `audio.defaultMicrophone` & `audio.defaultSpeaker` are populated.

**Comment by kotontrion** [comment]
Wp.get_default() returns as soon as the connection to pipewire has been established (or failed). The devices are then registered at a later point. Therefore these arrays are empty initially. So you have to connect to the corresponding signals. This is fine, as you usually want to update your widgets to reflect the current state and therefore have to connect to them anyway.

**Comment by adminy** [comment]
Hey @kotontrion , thanks for the tip. Is this really the solution though, feels like overengineering it. Especially the reactive part, since you kind of have no state on what goes on first, you have to prepare for everything ... It just becomes entangled web pretty quickly.
```typescript
function Menu() {
    const items: any = {}
    const menu = new Gtk.Menu()    
    const resize = () => menu.set_size_request(400, menu.get_size_request()[1]) 
    const added = (name: string) => {
        const item = items[name] = new Gtk.MenuItem({ label: name, visible: true })
        item.connect('activate', (source) => {
            // console.log('selected', source.label)
            for (const key in items) {
                items[key].deselect()
            }
            source.select()
            // TODO: actually change audio source
        })
        menu.append(items[name])
        resize()
    } 
    const removed = (name: string) => {
        if (!items[name]) return console.error('removing non-existent')
        menu.remove(items[name])
        delete items[name]
        resize()
    }
    const select = (name: string) => items[name]?.select() || name
    return { menu, added, removed, select }
}

function AudioDevices() {
    const audio = Wp.get_default()!.audio
    const mic = Menu()
    const speaker = Menu()
    const audioEvents = [
        'microphone-added',
        'microphone-removed',
        'speaker-added',
        'speaker-removed'
    ]
    const deviceTypes: any = { microphone: mic, speaker }
    for (const event of audioEvents) {
        audio.connect(event, (source, endpoint) => {
            const [type, action] = event.split('-')
            deviceTypes[type][action](endpoint.description)
            // console.log(type, action, endpoint.description)
        })
    }

    return <box className="AudioSlider">
        <button className="AudioButton"
            onDestroy={() => mic.menu?.destroy()}
            onClick={self => mic.menu?.popup_at_widget(self, Gdk.Gravity.SOUTH, Gdk.Gravity.NORTH, null)}
        >
            <box>
                <icon icon={bind(audio.defaultMicrophone, "volumeIcon").as(String)} />
                <label label={bind(audio.defaultMicrophone, "description").as(val => mic.select(val))} />
            </box>
        </button>
        <button className="AudioButton"
            onDestroy={() => speaker.menu?.destroy()}
            onClick={self => speaker.menu?.popup_at_widget(self, Gdk.Gravity.SOUTH, Gdk.Gravity.NORTH, null)}
        >
            <box>
                <icon icon={bind(audio.defaultSpeaker, "volumeIcon").as(String)} />
                <label label={bind(audio.defaultSpeaker, "description").as(val => speaker.select(val))} />
            </box>
        </button>

        <slider
            hexpand
            onDragged={({ value }) => audio.defaultSpeaker.volume = value}
            value={bind(audio.defaultSpeaker, "volume")}
        />
    </box>
}
```
and by using a gtk menu, I lost the ability to throw css at it.

**Comment by kotontrion** [comment]
Yes, generally all astal services are intended to be used from binds, or connections.

I quickly cleaned up your code to use binds instead of connections.

```js
import { bind,Gtk, Gdk, Widget } from "astal";
import Wp from "gi://AstalWp"

const AstalMenu = Widget.astalify(Gtk.Menu)
const AstalMenuItem = Widget.astalify(Gtk.MenuItem)

function MenuItem(device: Wp.Endpoint) {

  return AstalMenuItem({
        label: device.description,
        visible: true,
        onActivate: () => device.is_default = true
  })
}

function Menu(propname: keyof Wp.Audio) {
    const menu = <AstalMenu>
        {bind(Wp.get_default()!.audio, propname).as(sps => sps.map(MenuItem))}
    </AstalMenu>
    return menu
}

function AudioButton({type} : {type : "mic" | "speaker"}) {
    const { audio } = Wp.get_default()!;
    const menu = Menu(type === "mic" ? "microphones" : "speakers");
    const device = type === "mic" ? audio.defaultMicrophone : audio.defaultSpeaker;
    return <button className="AudioButton"
            onDestroy={() => menu.destroy()}
            onClick={self => menu.popup_at_widget(self, Gdk.Gravity.SOUTH, Gdk.Gravity.NORTH, null)}
        >
            <box>
                <icon icon={bind(device, "volumeIcon")} />
                <label label={bind(device, "description").as(String)} />
            </box>
        </button>

}

function AudioDevices() {

    const { audio } = Wp.get_default()!;
    
    return <box className="AudioSlider">
        <AudioButton type={"mic"} />
        <AudioButton type={"speaker"} />
        <slider
            hexpand
            onDragged={({ value }) => audio.defaultSpeaker.volume = value}
            value={bind(audio.defaultSpeaker, "volume")}
        />
    </box>
}

```

note that i discovered a bug which results in the bind not removing the menuitems of the menu. This is addressed in #25.

> and by using a gtk menu, I lost the ability to throw css at it.

you can add a className to the menu widget and refer that in css or use gtk widget selectors directly: 
```css
menu {
...
}

menuitem {
...
}
```

**Comment by adminy** [comment]
Thank you for opening my eyes on how to work with astal @kotontrion.
MenuItem did not accept being a reactive element, so the list is duplicated twice now.
```typescript
{bind(Wp.get_default()!.audio, propname)
   .as(sps => sps.map(MenuItem))}
```
I wonder if its something to do with bind only binding to properties not actual elements?

Its complaining about `(gjs:51011): astal-CRITICAL **: 09:32:27.605: could not set property "child" on [object instance wrapper GIName:Gtk.Menu jsobj@0x5cd0d9c8268 native@0xcf719c0]`

### Update

I managed to fix it with:
```typescript

import { bind, Gtk, Gdk, Widget, Binding } from "astal"

const GtkMenu = Widget.astalify(Gtk.Menu)
const AstalMenuItem = Widget.astalify(Gtk.MenuItem)

function AstalMenu({ child }: { child: Binding }) {
  return GtkMenu({ setup: self => self.hook(child, (menu, items) => {
    for (const item of menu.get_children()) {
      menu.remove(item)
    }
    for (const item of items) {
      menu.append(item)
    }
  })})
}

```
